File: sih/Frontend/src/types/contract.ts
/**
 * TypeScript mirror of API_CONTRACT.md - Sections 1, 2, 3 and 6.
 *
 * THIS FILE IS A MIRROR, NOT A SOURCE. `API_CONTRACT.md` is the single source of truth and is
 * copied byte-identically into all four domain folders. If a shape has to change, change it there
 * first, propagate to all four copies in the same commit, and tell the Backend owner.
 */

// -- Section 6: shared enums ---------------------------------------------------------------

export const BEHAVIOR_CLASSES = [
  "fixed",
  "periodic",
  "agile",
  "random",
  "intermittent",
] as const;
export type BehaviorClass = (typeof BEHAVIOR_CLASSES)[number];

/** `baseline` and `random` are the two open-loop references; the rest are learned. */
export const POLICY_TYPES = [
  "baseline",
  "random",
  "bandit",
  "q_learning",
  "dqn",
  "ppo",
  "index",
  "ctmc",
] as const;
export type PolicyType = (typeof POLICY_TYPES)[number];

export const SCENARIO_IDS = ["A", "B", "C", "D", "E", "F", "G"] as const;
export type ScenarioId = (typeof SCENARIO_IDS)[number];

export type DetectionType = "TP" | "FN" | "FP" | "TN";

export type SimulationStatus =
  | "draft"
  | "running"
  | "stopped"
  | "completed"
  | "failed";

// -- Section 1: standard envelope -----------------------------------------------------------

export interface ApiError {
  code: string;
  message: string;
  details: Record<string, unknown>;
}

export interface ApiEnvelope<T> {
  success: boolean;
  data?: T;
  error?: ApiError;
  requestId: string;
}

// -- Section 2: public resources -------------------------------------------------------------

export interface Simulation {
  id: string;
  name: string;
  status: SimulationStatus;
  bands: number;
  duration_steps: number;
  current_step: number;
  seed: number;
  policy_type: PolicyType;
  scenario_id: ScenarioId | null;
  created_at: string;
  live?: LiveSnapshot;
}

export interface CreateSimulationRequest {
  name: string;
  bands: number;
  durationSteps: number;
  seed?: number;
  scenario?: ScenarioId;
  policy?: PolicyType;
}

export interface Paged<T> {
  items: T[];
  page: number;
  size: number;
  total: number;
}

/** Shared by `GET /metrics/live` and the WebSocket reconnect replay - deliberately one shape. */
export interface LiveSnapshot {
  simulation_id: string;
  t: number;
  counters: { detections: number; false_alarms: number; reward: number };
  last_action?: number;
  scanned_bands?: number[];
  active_bands?: number[];
  detected_bands?: number[];
  reward?: number;
  model_id?: string;
  metrics?: MetricsSummary;
  error?: string;
}

export interface MetricsSummary {
  pd: number;
  pfa: number;
  ait: number;
  /**
   * AIT over every activation run, charging undetected ones the full episode.
   *
   * Use this to compare two policies. Raw `ait` only averages the runs a policy actually caught,
   * so a policy that intercepts more runs reports a worse `ait` - the extra runs it caught are
   * the hard, late ones the weaker policy missed entirely.
   */
  ait_censored: number;
  hpdr: number;
  median_latency: number;
  interception_ratio: number;
  scan_efficiency: number;
  precision: number;
  recall: number;
  f1: number;
  coverage: number;
  miss_rate: number;
  run_intercept_rate: number;
  detected_runs: number;
  total_runs: number;
  steps: number;
  total_scans: number;
  useful_scans: number;
  counts: { tp: number; fn: number; fp: number; tn: number };
  [key: string]: unknown;
}

export interface Scenario {
  id: ScenarioId;
  name: string;
  bands: number;
  emitters: number;
  duration_steps: number;
  episodes: number;
  emitter_mix: Record<BehaviorClass, number>;
  expected_outcome: string;
}

export interface Experiment {
  id: string;
  name: string;
  scenario: ScenarioId;
  policies: string[];
  episodes: number;
  seed: number;
  status: string;
  expected_outcome: string;
  created_at: string;
  progress?: ExperimentProgress;
}

export interface ExperimentProgress {
  fraction: number;
  completed_runs: number;
  total_runs: number;
  current_policy: string;
  current_episode: number;
  episodes_per_policy: number;
  finished: boolean;
  cancelled: boolean;
}

export interface MetricDelta {
  value: number;
  reference: number;
  absolute: number;
  percent: number | null;
}

export interface ExperimentResults {
  scenario: ScenarioId;
  scenario_name: string;
  episodes: number;
  duration_steps: number;
  seeds: number[];
  expected_outcome: string;
  policies: Record<string, MetricsSummary & { episodes: number }>;
  comparison: Record<string, Record<string, MetricDelta> | string>;
}

export interface SchedulerStatus {
  simulation_id: string;
  policy: PolicyType;
  status: string;
  step_count: number;
  ml_scheduler_degraded: boolean;
  ml_periodicity_degraded: boolean;
  available_policies: PolicyType[];
}

export interface ReadyStatus {
  status: "ready" | "degraded";
  ml_scheduler: "up" | "down";
  ml_periodicity: "up" | "down";
  running_simulations: number;
}

export interface ModelMetadata {
  model_id: string;
  algorithm: string;
  version: number;
  active: boolean;
  scenario: string | null;
  created_at: string;
  metrics: Record<string, unknown>;
  hyperparams: Record<string, unknown>;
  /** [min, max] training seed range, or null when the algorithm has nothing to train (e.g.
   * ctmc's fixed-form policy). Always present on the wire (Ai-ml-1's ModelRegistry.register
   * writes it on every entry) - missing here before 17 Sep was a stale mirror, not an
   * optional field the Backend might omit. */
  seed_range: [number, number] | null;
}

// -- Section 3: WebSocket frames ---------------------------------------------------------------

export type WsEventType =
  | "connection_ack"
  | "spectrum_update"
  | "scan_decision"
  | "detection_event"
  | "metrics_update"
  | "training_progress"
  | "error";

export interface WsFrame {
  type: WsEventType;
  simulation_id?: string;
  t?: number;
  active_bands?: number[];
  scanned_bands?: number[];
  detected_bands?: number[];
  false_alarm_bands?: number[];
  action?: number;
  model_id?: string;
  reward?: number;
  reward_terms?: Record<string, number>;
  valid_observation?: boolean;
  metrics?: MetricsSummary;
  status?: string;
  degraded?: boolean;
  snapshot?: LiveSnapshot;
  channels?: string[];
  heartbeat_seconds?: number;
  code?: string;
  message?: string;
  [key: string]: unknown;
}

export type WsChannel = "spectrum" | "scheduler" | "metrics" | "training";
export const WS_CHANNELS: WsChannel[] = [
  "spectrum",
  "scheduler",
  "metrics",
  "training",
];

File: sih/.claude/worktrees/sih-peptide-ppt-research-21209d/Ai-ml-1-Scheduler-Engine/ml/contract.py
"""Pydantic models for API_CONTRACT.md -- Sections 1, 4 and 6.

THIS FILE IS A MIRROR, NOT A SOURCE. ``API_CONTRACT.md`` is the single source of truth and is
copied byte-identically into all four domain folders. If a field has to change, change it in
``API_CONTRACT.md`` first, propagate it to all four copies in the same commit, and tell the
Backend owner -- do not let this file drift ahead of the contract, or the Backend will send a
StateVector this service silently mis-reads.

Field names, types and enum values below are transcribed from the contract verbatim.
"""

from __future__ import annotations

from typing import Any, Literal, Optional
from uuid import uuid4

from pydantic import BaseModel, ConfigDict, Field

# -- Section 6: shared enums and ID formats ------------------------------------------------

BehaviorClass = Literal["fixed", "periodic", "agile", "random", "intermittent"]
PolicyType = Literal["baseline", "bandit", "q_learning", "dqn", "ppo"]
DetectionType = Literal["TP", "FN", "FP", "TN"]
ScenarioId = Literal["A", "B", "C", "D", "E", "F", "G"]
LearningPolicy = Literal["bandit", "q_learning", "dqn", "ppo"]

SIMULATION_ID_PATTERN = r"^sim_[0-9a-f]{8}$"
EXPERIMENT_ID_PATTERN = r"^exp_[0-9a-f]{8}$"
MODEL_ID_PATTERN = r"^model_(baseline|bandit|q_learning|dqn|ppo)_[0-9a-f]{8}$"


def new_request_id() -> str:
    return f"req_{uuid4().hex[:8]}"


def new_model_id(algorithm: str) -> str:
    return f"model_{algorithm}_{uuid4().hex[:8]}"


def new_decision_id() -> str:
    return f"dec_{uuid4().hex[:8]}"


# -- Section 1: standard envelope ----------------------------------------------------------

class ErrorBody(BaseModel):
    code: str
    message: str
    details: dict[str, Any] = Field(default_factory=dict)


class SuccessEnvelope(BaseModel):
    success: Literal[True] = True
    data: Any = None
    requestId: str = Field(default_factory=new_request_id)


class ErrorEnvelope(BaseModel):
    success: Literal[False] = False
    error: ErrorBody
    requestId: str = Field(default_factory=new_request_id)


# -- Section 4: StateVector ----------------------------------------------------------------

class BandState(BaseModel):
    """One band's ML-001 features.

    ``periodicity_phase`` and ``periodicity_confidence`` are populated by the Backend from
    Ai-ml-2 before the StateVector reaches us. This service treats them as opaque inputs and
    never computes them (Ai-ml-1 README, "State / Action / Reward").
    """

    model_config = ConfigDict(extra="forbid")

    band_id: int = Field(ge=0)
    time_since_last_scan: int = Field(ge=0)
    recent_detection_rate_ewma: float = Field(ge=0.0, le=1.0)
    consecutive_misses: int = Field(ge=0)
    periodicity_phase: float = Field(ge=0.0, le=1.0)
    periodicity_confidence: float = Field(ge=0.0, le=1.0)
    band_priority_weight: float = Field(ge=0.0)
    tuning_cost_to_band: int = Field(ge=0)


class ReceiverState(BaseModel):
    model_config = ConfigDict(extra="forbid")

    tuned_bands: list[int] = Field(default_factory=list)
    dwell_remaining_ms: int = Field(default=0, ge=0)
    tuning_delay_countdown_ms: int = Field(default=0, ge=0)


class StateVector(BaseModel):
    model_config = ConfigDict(extra="forbid")

    bands: list[BandState] = Field(min_length=1)
    receiver: ReceiverState = Field(default_factory=ReceiverState)


class Action(BaseModel):
    model_config = ConfigDict(extra="forbid")

    next_band: int = Field(ge=0)
    dwell_time: Optional[int] = Field(default=None, ge=0)


# -- Section 4: request/response bodies ----------------------------------------------------

class DecideRequest(BaseModel):
    simulation_id: str
    state: StateVector
    policy: LearningPolicy
    model_id: Optional[str] = None


class DecideResponse(BaseModel):
    action: Action
    model_id: str
    decision_id: str


class LearnRequest(BaseModel):
    simulation_id: str
    decision_id: str
    state: StateVector
    action: Action
    reward: float
    next_state: Optional[StateVector] = None


class LearnResponse(BaseModel):
    acknowledged: bool = True


class TrainRequest(BaseModel):
    algorithm: LearningPolicy
    scenario: ScenarioId
    hyperparams: dict[str, Any] = Field(default_factory=dict)
    episode_count: Optional[int] = Field(default=None, ge=1)
    seed_range: Optional[tuple[int, int]] = None


class TrainResponse(BaseModel):
    job_id: str


class TrainStatusResponse(BaseModel):
    status: Literal["running", "done", "failed"]
    progress: float = Field(ge=0.0, le=1.0)
    detail: dict[str, Any] = Field(default_factory=dict)


class EvaluateRequest(BaseModel):
    scenario: ScenarioId
    episode_count: int = Field(default=20, ge=1)


class ModelMetadata(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    model_id: str
    algorithm: PolicyType
    scenario: Optional[str] = None
    version: int = 1
    active: bool = False
    created_at: str
    hyperparams: dict[str, Any] = Field(default_factory=dict)
    seed_range: Optional[list[int]] = None
    metrics: dict[str, Any] = Field(default_factory=dict)


class HealthResponse(BaseModel):
    status: Literal["ok"] = "ok"
